Added function to create `std::variant` for multiple array views.

odlomax opened the pull request:
During PR #218, I found myself writing two nested switch statements in order to create an `ArrayView` from an `Array`. I've found myself writing this same pattern multiple times throughout Atlas, and indeed other external applications which use atlas.

This PR adds provides an alternative to the nested switch statement. The method `util::make_array_view_variant<Values<types...>, Ranks<ints...>>(Array& array)` creates an `std::variant` of `ArrayView` types, and then assigns the view that matches `array` (so long as the value-type and rank can be found within the `Values` and `Ranks` arguments respectively. The `ArrayView` can then be accessed rather succinctly using `std::visit`.

wdeconinck commented:
Hi @odlomax I think this is a great idea. 
I was wondering if this should not become part of the "array" namespace. If you want we could discuss offline.
It should also cater for device views though.
`array::make_host_view_variant` and `array::make_device_view_variant`
All possibly supported Values and Ranks could be immediately part of the variant type, removing even more boiler-plate?

odlomax commented:
> Hi @odlomax I think this is a great idea. I was wondering if this should not become part of the "array" namespace. If you want we could discuss offline. It should also cater for device views though. `array::make_host_view_variant` and `array::make_device_view_variant` All possibly supported Values and Ranks could be immediately part of the variant type, removing even more boiler-plate?

Hi @wdeconinck,

Thanks for the feedback! I'm on leave at the moment, but this is oddly recreational between baby's naps. I'll arrange a catch up when I'm back after next week. 🙂

odlomax commented:
Hi @wdeconinck. I believe I've incorporated all of your suggestions. It's ready for a look when you have the time. Cheers!

odlomax commented:
@wdeconinck Apologies for the last minute push.

There were relatively important EXPECT statements that needed to be added to lines [139](https://github.com/ecmwf/atlas/pull/220/files#diff-f695f80bd1611cbb0f696febb67f02415a51b5b07c3cde831790ef961ab84e82R139) and [164](https://github.com/ecmwf/atlas/pull/220/files#diff-f695f80bd1611cbb0f696febb67f02415a51b5b07c3cde831790ef961ab84e82R164) of the test. I also removed some extraneous whitespace at the end of an otherwise untouched file.

odlomax commented:
Okay, I won't touch this until it's reviewed. I promise!

odlomax commented on src/atlas/array/ArrayViewVariant.h:
```diff
@@ -0,0 +1,79 @@
+
+/// @brief Variant containing all supported ArrayView alternatives.
+using ArrayViewVariant = detail::Variant<Values, Ranks>;
+
```
In case you're interested, `array::ArrayViewVariant` currently expands into:

```
std::variant<atlas::array::ArrayView<float const, 1>,
             atlas::array::ArrayView<float const, 2>,
             atlas::array::ArrayView<float const, 3>,
             atlas::array::ArrayView<float const, 4>,
             atlas::array::ArrayView<float const, 5>,
             atlas::array::ArrayView<float const, 6>,
             atlas::array::ArrayView<float const, 7>,
             atlas::array::ArrayView<float const, 8>,
             atlas::array::ArrayView<float const, 9>,
             atlas::array::ArrayView<float, 1>,
             atlas::array::ArrayView<float, 2>,
             atlas::array::ArrayView<float, 3>,
             atlas::array::ArrayView<float, 4>,
             atlas::array::ArrayView<float, 5>,
             atlas::array::ArrayView<float, 6>,
             atlas::array::ArrayView<float, 7>,
             atlas::array::ArrayView<float, 8>,
             atlas::array::ArrayView<float, 9>,
             atlas::array::ArrayView<double const, 1>,
             atlas::array::ArrayView<double const, 2>,
             atlas::array::ArrayView<double const, 3>,
             atlas::array::ArrayView<double const, 4>,
             atlas::array::ArrayView<double const, 5>,
             atlas::array::ArrayView<double const, 6>,
             atlas::array::ArrayView<double const, 7>,
             atlas::array::ArrayView<double const, 8>,
             atlas::array::ArrayView<double const, 9>,
             atlas::array::ArrayView<double, 1>,
             atlas::array::ArrayView<double, 2>,
             atlas::array::ArrayView<double, 3>,
             atlas::array::ArrayView<double, 4>,
             atlas::array::ArrayView<double, 5>,
             atlas::array::ArrayView<double, 6>,
             atlas::array::ArrayView<double, 7>,
             atlas::array::ArrayView<double, 8>,
             atlas::array::ArrayView<double, 9>,
             atlas::array::ArrayView<int const, 1>,
             atlas::array::ArrayView<int const, 2>,
             atlas::array::ArrayView<int const, 3>,
             atlas::array::ArrayView<int const, 4>,
             atlas::array::ArrayView<int const, 5>,
             atlas::array::ArrayView<int const, 6>,
             atlas::array::ArrayView<int const, 7>,
             atlas::array::ArrayView<int const, 8>,
             atlas::array::ArrayView<int const, 9>,
             atlas::array::ArrayView<int, 1>,
             atlas::array::ArrayView<int, 2>,
             atlas::array::ArrayView<int, 3>,
             atlas::array::ArrayView<int, 4>,
             atlas::array::ArrayView<int, 5>,
             atlas::array::ArrayView<int, 6>,
             atlas::array::ArrayView<int, 7>,
             atlas::array::ArrayView<int, 8>,
             atlas::array::ArrayView<int, 9>,
             atlas::array::ArrayView<long const, 1>,
             atlas::array::ArrayView<long const, 2>,
             atlas::array::ArrayView<long const, 3>,
             atlas::array::ArrayView<long const, 4>,
             atlas::array::ArrayView<long const, 5>,
             atlas::array::ArrayView<long const, 6>,
             atlas::array::ArrayView<long const, 7>,
             atlas::array::ArrayView<long const, 8>,
             atlas::array::ArrayView<long const, 9>,
             atlas::array::ArrayView<long, 1>,
             atlas::array::ArrayView<long, 2>,
             atlas::array::ArrayView<long, 3>,
             atlas::array::ArrayView<long, 4>,
             atlas::array::ArrayView<long, 5>,
             atlas::array::ArrayView<long, 6>,
             atlas::array::ArrayView<long, 7>,
             atlas::array::ArrayView<long, 8>,
             atlas::array::ArrayView<long, 9>,
             atlas::array::ArrayView<unsigned long const, 1>,
             atlas::array::ArrayView<unsigned long const, 2>,
             atlas::array::ArrayView<unsigned long const, 3>,
             atlas::array::ArrayView<unsigned long const, 4>,
             atlas::array::ArrayView<unsigned long const, 5>,
             atlas::array::ArrayView<unsigned long const, 6>,
             atlas::array::ArrayView<unsigned long const, 7>,
             atlas::array::ArrayView<unsigned long const, 8>,
             atlas::array::ArrayView<unsigned long const, 9>,
             atlas::array::ArrayView<unsigned long, 1>,
             atlas::array::ArrayView<unsigned long, 2>,
             atlas::array::ArrayView<unsigned long, 3>,
             atlas::array::ArrayView<unsigned long, 4>,
             atlas::array::ArrayView<unsigned long, 5>,
             atlas::array::ArrayView<unsigned long, 6>,
             atlas::array::ArrayView<unsigned long, 7>,
             atlas::array::ArrayView<unsigned long, 8>,
             atlas::array::ArrayView<unsigned long, 9>>;
```

odlomax replied:
This has now been split into const and non-const variants!

odlomax commented:
Hi @wdeconinck,

When attempting to using this in anger (see PR #226), I found that it was worth refactoring the `make_view_variant` methods to produce either `const` or `non-const` versions of the variant depending on the constness of the array. Shall I merge the refactor into this PR, or leave it in #226?

odlomax commented:
Hi @wdeconinck  @sbrdar,

I've refactored the introspection helpers as we discussed on Slack. I've also added an additional helper to distinguish between `value_type` and `non_const_value_type`. Seemed like a sensible idea.

Cheers!

odlomax commented:
@wdeconinck the latest push has your suggested signature for the introspection helpers. Thanks for the suggestion, it's a lot tidier.

odlomax commented:
Basic GHA is passing now!

https://github.com/JCSDA-internal/atlas/actions/runs/11157940876

odlomax commented:
MacOS tests! Nooo!

wdeconinck commented:
> MacOS tests! Nooo!

https://github.com/ecmwf/atlas/actions/runs/11157941275/job/31068813417?pr=220 should not be the problem of this PR, but about the CI infrastructure. I passed on this problem.

wdeconinck commented on src/atlas/array/ArrayViewVariant.h:
```diff
@@ -0,0 +1,102 @@
+}  // namespace detail
+
+/// @brief Supported ArrayView value types.
+using ValueTypes = detail::Types<float, double, int, long, unsigned long>;
+
+/// @brief Supported ArrayView ranks.
+using Ranks = detail::Ints<1, 2, 3, 4, 5, 6, 7, 8, 9>;
```
I would prefer if these were in the detail namespace as well, or is this something you require access to in user code?

odlomax replied:
Yeah, probably an oversight on my part. Makes more sense to have them _reasonably_ hidden in the  `detail` namespace.

odlomax replied:
Done!

wdeconinck reviewed (approved):
This looks great now! Thanks!
